packages/react/src/internal/useControllableState.js

```diff
@@ -8,68 +8,90 @@
 import { useEffect, useRef, useState } from 'react';
 import { warning } from './warning';
 
-export function useControllableState(
-  controlledState,
-  controlledSetState,
-  defaultValue
-) {
-  const controlled = useRef(controlledState !== undefined);
-  const [state, internalSetState] = useState(() => {
-    if (controlled.current === true) {
-      return controlledState;
-    }
-    return defaultValue;
-  });
+/**
+ * This custom hook simplifies the behavior of a component if it has state that
+ * can be both controlled and uncontrolled. It functions identical to a
+ * useState() hook and provides [state, setState] for you to use. You can use
+ * the `onChange` argument to allow updates to the `state` to be communicated to
+ * owners of controlled components.
+ *
+ * Note: this hook will warn if a component is switching from controlled to
+ * uncontrolled, or vice-verse.
+ *
+ * @param {object} config
+ * @param {string} config.name - the name of the custom component
+ * @param {any} config.defaultValue - the default value used for the state. This will be
+ * the fallback value used if `value` is not defined.
+ * @param {Function} config.onChange - an optional function that is called when
+ * the value of the state changes. This is useful for communicating to parents of
+ * controlled components that the value is requesting to be changed.
+ * @param {any} config.value - a controlled value. Omitting this means that the state is
+ * uncontrolled
+ * @returns {[any, Function]}
+ */
+export function useControllableState({
+  defaultValue,
+  name = 'custom',
```

Comment:
Out of curiosity, what would an example of `name` being used be? Would it be the name of the state being used or the component that is being controlled?

Reply:
@abbeyhrt the idea with `name` was that a component could provide the `name` argument to offer better development error messages. So, for example, `Tabs` could `useControllableState({ name: 'Tabs' })` and it would say that "a Tabs component is switching from controlled to uncontrolled"